File: hangman-game/new/Hangman.java
import java.util.List;
import java.util.ArrayList;
import java.util.Scanner;
import java.lang.Math;


class Generator {
    private static final List<String> WORDS = new ArrayList<>(List.of("Harry", "Hermione", "Ron"));
    
    static String generate() {
        int min = 0;
        int max = Generator.WORDS.size() - 1;
        int range = max - min;
        int randomIndex =  (int)(Math.random() * range) + min;
        return Generator.WORDS.get(randomIndex);
    }
}

class Game {
    private int tries;
    private String solution;
    private String[] word;

    Game() {
        this.tries = 0;
        this.solution = Generator.generate().toLowerCase();
        this.word = new String[this.solution.length()];
        for (int i = 0; i < this.solution.length(); i++) {
            this.word[i] = "_";
        }
    }

    void displayWord() {
        System.out.println("The word is " + String.join(" ", this.word) + ".");
    }

    void replaceWord(final String guess, final List<Integer> indices) {
        for (Integer index : indices) {
            this.word[index] = guess;
        }
    }

    void solve() {
        for (int i = 0; i < this.word.length; i++) {
            this.word[i] = this.solution.substring(i, i+1);
        }
    }

    void increaseTries() {
        if (this.tries < Player.getMaxTries()) {
            tries++;
        }
    }

    boolean isSolved() {
        for (int i = 0; i < this.solution.length(); i++) {
            if (! this.word[i].equals(this.solution.substring(i, i+1))) {
                return false;
            }
        }
        return true;
    }

    int getTries() {
        return this.tries;
    }

    String getSolution() {
        return this.solution;
    }
}

class Checker {

    static List<Integer> check(final String guess, final String solution) {
        if (guess.length() == 1) {
            return Checker.checkGuess(guess.charAt(0), solution);
        } else {
            return Checker.checkGuess(guess, solution);
        }

    }

    static List<Integer> checkGuess(final char guess, final String solution) {
        List<Integer> indices = new ArrayList<>();
        for (int i = 0; i < solution.length(); i++) {
            if (solution.substring(i,i+1).equals(String.valueOf(guess))) {
                indices.add(i);
                continue;
            }
        }
        return indices;
    }

    static List<Integer> checkGuess(final String guess, final String solution) {
        List<Integer> bool = new ArrayList<>();
        if (guess.equals(solution)) {
            bool.add("true".hashCode());
        } else {
            bool.add("false".hashCode());
        }
        return bool;
    }
}

class Player {
    private Game game;
    private static final int MAX_TRIES = 8;

    Player() {
        this.game = new Game();
    }

    String guess() {
        System.out.print("Enter a letter:\n>>> ");
        Scanner scanner = new Scanner(System.in);
        String input = scanner.nextLine();
        return input;
    }

    boolean isAlive() {
        if (this.game.getTries() < Player.MAX_TRIES) {
            return true;
        }
        return false;
    }

    void win(final int lives) {
        System.out.println("You won with " + String.valueOf(lives) + " tries remaining!");
    }

    void lose() {
        System.out.println("You lose!");
    }

    void play() {
        while (this.isAlive()) {
            System.out.println(this.game.getSolution());
            this.game.displayWord();
            String guess = this.guess();
            List<Integer> result = Checker.check(guess, this.game.getSolution());
            if (result.size() == 0 || result.get(0) == 97196323) {
                // Try again
                this.game.increaseTries();
                continue;
            } else if (result.get(0) == 3569038) {
                // Instant win
                this.game.solve();
                this.win(this.MAX_TRIES - this.game.getTries());
                break;
            } else {
                this.game.replaceWord(guess, result);
                if (this.game.isSolved()) {
                    this.win(this.MAX_TRIES - this.game.getTries());
                    break;
                }
            }
        }
        if (! this.game.isSolved()) {
            this.lose();
        }
    }

    static int getMaxTries() {
        return Player.MAX_TRIES;
    }
}

class Hangman {
    private static boolean play = true;

    static boolean ask() {
        System.out.print("Do you want to play a game? [y/n]\n>>>");
        Scanner scanner = new Scanner(System.in);
        String input = scanner.nextLine().substring(0,1).toLowerCase();
        if (input.equals("y")) {
            return true;
        }
        return false;
    }

    static void loop() {
        if (Hangman.ask()) {
            Player player = new Player();
            player.play();
            return;
        }
        Hangman.play = false;
    }
    public static void main(String[] args) {
        while (Hangman.play) {
            Hangman.loop();
        }
    }
}

File: hangman-game/new/hangman.py
#!/usr/bin/env python3
"""
    Author: Pek, Jonetta
    Date: 6 October 2023
    Purpose: Single-player word guessing game with limited number of tries
"""


import random


class Checker():

    def __init__(self):
        pass

    
    def check_guess(self, guess: str, word: str) -> [list | bool]:
        if guess in word:
            return [index for index in range(len(word)) if word[index] == guess]
        return False


    def check_word(self, guess: str, word: str) -> bool:
        return True if guess == word else False
    

class Game():
    __WORDS = ['Harry', 'Hermione', 'Ron', 'Dobby', 'Hagrid', 'Dumbledore', 'Minerva', 'Sirius']

    def __init__(self):
        self.__num_guess: int = 0
        self.__solution: str = self.__WORDS[random.randint(0,len(self.__WORDS)-1)].lower()
        self.__word: list = ['_'for _ in self.__solution]


    def display_word(self) -> None:
        print(f'The word is {" ".join(self.__word)}.')


    def replace_word(self, guess: str, indices: [list | bool]) -> None:
        if indices:
            for index in indices:
                self.__word[index] = guess


    def increase_count(self):
        if self.__num_guess < Player.MAX_GUESS():
            self.__num_guess += 1


    def is_solved(self) -> bool:
        return ''.join(self.__word) == self.__solution


    def instant_solve(self) -> None:
        self.__word = self.__solution.split()
        

    @property
    def num_guess(self):
        return self.__num_guess
    

    @property
    def solution(self):
        return self.__solution


class Player():
    __MAX_GUESS = 8

    def __init__(self):
        self.__game = Game()
        self.__checker = Checker()
    

    def guess(self) -> str:
        return input('Enter a letter:\n>>>').lower()


    def is_alive(self) -> bool:
        return True if self.__game.num_guess < Player.MAX_GUESS() else False
    

    def win(self) -> bool:
        print('You win!')
        return True


    def lose(self) -> bool:
        print('You lose!')
        return False


    def play(self):
        while self.is_alive():
            self.__game.display_word()
            guess: str = self.guess()
            # Instant win
            if len(guess) == len(self.__game.solution) and self.__checker.check_word(guess, self.__game.solution):
                self.__game.instant_solve()
                self.win()
                break
            is_present: [list | bool] = self.__checker.check_guess(guess, self.__game.solution)
            if is_present:
                self.__game.replace_word(guess, is_present)
            if not is_present:
                self.__game.increase_count()
            if self.__game.is_solved():
                self.win()
                break
        if not self.__game.is_solved():
            self.lose()

    @classmethod
    def MAX_GUESS(cls):
        return cls.__MAX_GUESS
        
def loop():
    ans = input('Would you like to play a game?(y/n):\n>>>')
    if ans[0].lower() == 'y':
        player = Player()
        player.play()
    else:
        return

if __name__ == '__main__':
    while True:
        loop()
